[drake] Vendingmachine step3

DrakeYang opened the pull request:
피드백 부탁드립니다.

DrakeYang committed:
자판기 객체 파일 생성

DrakeYang committed:
자판기 금액변수,조절함수 추가

DrakeYang committed:
금액 수정 함수명 변경.
insert > plus, use > minus

DrakeYang committed:
음료별 재고 배열 추가, 배열에 음료 추가하는 함수 추가

DrakeYang committed:
음료추가함수 메인 외 나머지 함수들 private 처리

DrakeYang committed:
자판기 객체 생성 후 음료 추가 까지 테스트 완료

DrakeYang committed:
자판기 객체파일의 대상으로 단위테스트 파일 추가

DrakeYang committed:
1,2단계 진행내용 정리해서 추가

DrakeYang committed:
음료객체를 받아서 배열에 추가하는 함수 테스트 추가

DrakeYang committed:
가격을 리턴하는 함수 추가

DrakeYang committed:
음료객체의 이름 리턴함수 추가

DrakeYang committed:
전체 재고 출력함수 추가

DrakeYang committed:
음료 추가후 확인 테스트 실패의 경우도 추가

DrakeYang committed:
종류별 음료수 배열의 마지막 팝하는 함수추가.
음료별 주문하는 함수 추가(팝->주문내역으로 이동).
금액 확인 함수 추가

DrakeYang committed:
자판기 재고 추가함수 추가

DrakeYang committed:
음료수 생성 및 생성된 음료 자판기로 추가

DrakeYang committed:
제로칼로리 콜라 추가.
음료수 각2개씩 추가하도록 수정

DrakeYang committed:
인풋뷰 클래스 추가

DrakeYang committed:
유저입력 받는 함수 추가.
자판기 시작메뉴 메세지 추가

DrakeYang committed:
아웃풋뷰 객체 추가

DrakeYang committed:
자판기 입력금액 출력 메세지 함수 추가

DrakeYang committed:
남아있는 재고 출력함수 추가

DrakeYang committed:
남아있는 재고 문자열로 리턴하는 함수 추가

DrakeYang committed:
재고 출력 메세지 수정

DrakeYang committed:
함수명 변경 startVendingMachine -> mainMenu

DrakeYang committed:
체커객체 추가

DrakeYang committed:
문자열을 받아서 양의 정수로 치환가능한지 체크하는 함수 추가

DrakeYang committed:
돈추가,1,2단계 메뉴 선택 함수 추가

DrakeYang committed:
잘못된 금액 입력시 에러출력 함수 추가

DrakeYang committed:
첫번째 메뉴와 두번째 메뉴 통합. 순서에 맞게 다른 함수들 조정

DrakeYang committed:
종료메뉴 및 메세지 추가

DrakeYang committed:
재고 출력하는 함수 리팩토링.
인벤토리 한개씩 리턴하는 함수와 전체리턴하는 함수로 분리

DrakeYang committed:
재고출력을 배열로 바꿈

DrakeYang committed:
음료선택 메뉴함수를 추가

DrakeYang committed:
음료재고출력함수 입력값을 배열로 수정

DrakeYang committed:
재고 출력을 위해 이름,가격을 리턴하는 프로토콜 추가.음료객체가 프로토콜 준수

DrakeYang committed:
음료들을 담는 배열객체 음료창고 를 파일로 분리

DrakeYang committed:
프로토콜을 걷어내고 재고정보만을 담는 객체 InventoryDetail 추가 및 적용

DrakeYang committed:
음료수 구매 성공메세지 추가

DrakeYang committed:
음료 재고 개수를 뜻하는 단어를 numver->count 로 변경

DrakeYang committed:
문자열을 받아서 숫자 1~maxDrinkNumber 사이인지 체크하는 함수 추가

DrakeYang committed:
수량을 뜻하는 영어단어 number->count 로 수정

DrakeYang committed:
음료 구매 단계 리팩토링 : 나눔.
음료 선택 메뉴 if문으로 수정.
선택 후 구매단계별로 함수로 나눔

DrakeYang committed:
재고창고->주문목록 으로 음료 이동하는 함수 추가.
추가된 함수 메인함수 순서에 적용

DrakeYang committed:
메뉴선택시 +1 번 적용되던 부분 수정

DrakeYang committed:
잔액부족 메세지 추가

DrakeYang committed:
주문금액과 잔액을 비교하는 함수 추가

DrakeYang committed:
리드미 추가

godrm commented on VendingMachine/VendingMachine/Drink.swift:
```diff
@@ -37,6 +37,13 @@ class Drink : CustomStringConvertible {
+    func getName() ->String {
+        return self.name
+    }
+    
```
Drink에서 price를 가져가서 어떻게 사용하는지 고민해 보는게 좋습니다.
항상 값을 그대로 주기보다는 기준값을 받아서 객체가 내부 값과 판단하도록 구현하는 게 좋습니다.
예를 들어 일정 금액이 넘는 객체를 찾아야 한다면, 
각 객체 price를 가져와서 비교하지말고, 기준 금액을 넘겨주면서 price가 더 적은 금액인지 판단하는 메소드를 호출하는거죠. 
덧붙여서 스위프트에서 getName() 처럼 (어쩔수없이) 값을 가져만 가는 경우는, private(set)을 활용하는게 더 깔끔합니다.

DrakeYang replied:
private(set) 은 var 에서만 쓸수 있어서 우선은 안쓰겠습니다.
가격비교는 데이터 전송용 객체인 InventoryDetail 에서 하면 되서 Drink 객체 내부에는 안써도 될것 같습니다.

godrm commented on VendingMachine/VendingMachine/DrinkInventory.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+    func getDrinkCount()->Int{
+        return self.drinkCount
+    }
```
InventoryDetail 처럼 로직이 없이 값만 넘기는 Value Object 또는 Data Transfer Object 객체는 그냥 public 으로 만드세요. 별도 getter 함수를 만들 필요 없습니다.

DrakeYang replied:
수정했습니다!

godrm commented on VendingMachine/VendingMachine/OutputView.swift:
```diff
@@ -0,0 +1,71 @@
+    /// 숫자를 입력해야되는데 숫자가 아닌경우
+    func notNumeric(){
+        print("정수가 아닙니다.")
+    }
```
출력용이더라도 이렇게 함수내에 하드 코딩(?)하면 안됩니다. 
enum 으로 선언을 하던지 해서 매개변수로 넘겨서 하나의 함수로 표현하세요!!

DrakeYang replied:
옙!

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,112 @@
+            return nil
+        }
+        orderedDrinks.append(drink)
+        return ()
```
`orderLowSugarChocoMilk() -> ()?` 이렇게 리턴타입을 의미없는 형태로 만든 이유는 뭔가요?
그 이후에 getOrederdDrinks()를 꼭 호출해야만 하는데, `[Drink]` 나 `[Drink]?` 이런 형태가 되면 어떨까요?

DrakeYang replied:
자판기 내부에서 돌아가는 함수라 리턴타입이 없는 함수인데 실패할 경우를 체크하고 싶었습니다. 
그냥 리턴값을 Bool 로 해서 성공/실패 를 결과로 표시하는게 나을까요? 
아니면 리턴값이 없는 함수는 없는게 나은가요?

godrm replied:
몇 가지 선택이 가능합니다. 
* 우선 정상적인 경우만 리턴하고 예외적인 경우는 throws 를 활용하는거죠.
* 아니면 미리 상황을 판단해서 return 하는 메소드를 사용하고, 동작하는 함수는 return 값이 없도록 만드는거죠
함수 하나로 여러가지는 판단하는건가 아닌가 고민해보고 결정하세요.

DrakeYang replied:
리턴값이 ()? 인 함수들 쪼갰습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,160 @@
+        }
+        
+        // 성공메세지 출력
+        outputView.buyingSuccessMessage(dirnkName: drinkDetail.getDrinkName(), drinkCount: orderCount, drinkPirce: drinkDetail.getDrinkPrice()*orderCount)
```
drinkDetail를 사용할꺼면 매개변수를 속성별로 나눠서 쓸 필요없이 drinkDetail 객체 인스턴스 자체를 넘기는게 좋습니다.
그리고 `drinkDetail.getDrinkPrice()*orderCount` 이렇게 쓰지 않도록 객체 내부에 메소드를 추가하는게 어떨까요?

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,160 @@
     
+    while true {        
+        let firstMenu = mainManu()
+        switch firstMenu {
```
mainMenu() (오타확인해주세요)
에서 리턴하는 값을 String 대신에 딱 경우의 수만 확인하도록 enum 으로 지정하세요.
그래야 switch-case 구문에서도 실수하지 않도록 할 수 있습니다.

DrakeYang replied:
switch case 부분 수정했습니다.

godrm commented on VendingMachine/VendingMachine/VendingMachine.swift:
```diff
@@ -0,0 +1,112 @@
+    }
+    
+    /// 남아있는 모든 재고 확인
+    func getAllInventory()->[InventoryDetail?]{
```
`[InventoryDetail?]` 로 선택한 이유가 있나요? 
딱 재고가 있는 메뉴만 `[InventoryDetail]`로 리턴해주는게 더 편리하지 않을까요? 
재고가 없는 음료가 nil로 있어야 할 필요가 있나요??

DrakeYang replied:
자판기에 매진된 음료는 빨간색으로 매진 이라고 뜨는것처럼 매진된걸 확인하고 싶었습니다.
매진된 음료는 출력되지 않도록 수정했습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,160 @@
+        case 3 : vendingMachine.orderCoke()
+        case 4 : vendingMachine.orderZeroCalorieCoke()
+        case 5 : vendingMachine.orderHotTopCoffee()
+        case 6 : vendingMachine.orderEnergyDrink()
```
여기도 drinkNumber 대신에 enum으로 개선해보세요. 메뉴선택할 때부터 Int, String 을 쓰기보다는 정해진 enum 값만 선택하도록 하세요.

DrakeYang replied:
옙

DrakeYang committed:
OutputView 의 메세지출력 함수들 print -> return 으로 수정

DrakeYang committed:
InventoryDetail getter 제거, 변수 private 제거

DrakeYang committed:
InventoryDetail getter 제거내용 적용

DrakeYang committed:
첫번째 메뉴 입력받는 방식 병경 string->enum.
InputView 에 FirstMenu 로 enum 선언. 유저 인풋을 입력받아서 Checker 의 checkFirstMenuInput 가 맞는지 체크.
체크 후 메인메뉴에서 받아서 진행

DrakeYang committed:
음료종류 선택을 위한 enum DrinkNumber 추가.
유저입력을 받아서 DrinkNumber 인지 체크하는 함수 추가

DrakeYang committed:
음료종류 선택하는 enum DrinkNumber 메인함수에 적용

DrakeYang committed:
whichDrink 함수 InputView > OutputView 으로 이동.
print -> return 으로 변경

DrakeYang committed:
음료추가에 관련된 함수들 리턴타임 ()? 모두 수정.
음료 추가시 재고정보 리턴하도록 수정. 음료추가 단계 세분화

DrakeYang committed:
buyingSuccessMessage 입력인자 dirnk > drink 오타수정

DrakeYang committed:
재고->주문목록 으로 이동하는 함수 리턴값 ()? > InventoryDetail 로 수정.
음료번호를 받아서 음료타입으로 변환하는 함수 추가.
재고정보를 받아서 음료이동 성공메세지를 출력하는 함수 추가.
음료이동에 관련된 함수들 메인함수에 적용

DrakeYang committed:
재고정보를 받는 음료이동성공메세지에서 drinkDetail.drinkCount * drinkDetail.drinkPrice)원 으로 수정

DrakeYang committed:
재고가 없는 음료는 목록에 출력되지 않도록 수정:

DrakeYang committed:
재고가 없는 음료는 목록에 출력되지 않도록 수정

DrakeYang committed:
라인정리

DrakeYang committed:
배운점들 리드미에 추가

DrakeYang commented:
피드백 반영했습니다!

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,183 @@
+            buyingDrink(drinkNumber: selectedDrinkNumber)
+        } // 잘못된 선택일 경우
+        else {
+            print(outputView.wrongMenu())
```
이 부분은 제 피드백을 다르게 반영한 것 같습니다.
* print() 구분은 outputView 내부에 있어야 합니다.
* outputView 에 출력하는 코드를 하드코딩 하지 말고 매개변수로 넘기라는 의미입니다. enum 타입을 고려하세요.
* 결과적으로 main에서는 outputView를 호출하면서 에러 메시지 종류를 매개변수로 넘기는게 됩니다.

DrakeYang replied:
에러메세지들 enum 으로 묶었습니다. print 담당함수를 추가했습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,183 @@
+    }
+    
+    /// 음료선택시 재고가 남아있는지 체크후 해당 음료의 정보를 가저온다
+    func checkInventoryCount(dirnkNumber:InputView.DrinkNumber)->InventoryDetail?{
```
dirnkNumber 변수명에 오타 확인해주세요.

DrakeYang replied:
옙

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,183 @@
+    func calculateMoney(drink:InventoryDetail,orderCount:Int)->()?{
+        if vendingMachine.getMoney() >= drink.drinkPrice*orderCount {
+            // 금액 차감
+            vendingMachine.minusMoney(money: drink.drinkPrice*orderCount)
```
이 동작도 main에서 구현하는 것보다는 vendingMachine 내부에 있는게 좋습니다.
getMoney() 대신 아니라 총 금액으로 구매 가능한지 판단하는 메소드가 있거나
minusMoney() 역할까지 대신할 수 있는 메소드로 처리가 가능합니다.
이렇게 객체 내부에 값을 가져와서 비교, 계산, 처리하기 보다는 기준값을 넘겨주도록 개선해보세요.

DrakeYang replied:
내부에서 처리후 결과를 넘겨서 출력함수가 받는방식으로 수정했습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,183 @@
+        /// 구매가 가능한지 체크한다
+        guard let drinkDetail = checkInventoryCount(dirnkNumber: drinkNumber) // 구매하려는 음료가 잔고가 있는지
+            , let orderCount = getOrderCount(drinkName: drinkDetail.drinkName) // 원하는 수량이 >0 인지
+            , checkEnoughDrinkCount(drinkCount: drinkDetail.drinkCount, orderCount: orderCount) == true
```
checkInventoryCount() , getOrderCount() , checkEnoughDrinkCount() 어떤건 값을 가져오는건지, 있는지 없는지 판단하는건지 헷갈리네요. API 디자인 가이드라인에 따르면 Bool 값인 경우는 has-, is- 등으로 명확하게 구분하는게 좋습니다.

DrakeYang replied:
수정했습니다.

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,183 @@
+        } else {
+            // 이동 실패시
+            print(outputView.notEnoughDrink())
+        }
```
이 부분도 여기서 출력하기보다는 orderDrinks()를 호출하고 결과를 outputView로 넘겨서 해당하는 결과값에 따라서 에러 메시지를 출력하도록 개선해보세요.

DrakeYang committed:
문자열을 받아서 프린트 하는 함수 추가.
에러메제시 enum 으로 정리

DrakeYang committed:
enum 오타 수정

DrakeYang committed:
에러메세지 출력방식 enum 으로 변경된 함수 적용

DrakeYang committed:
오타수정 : dirnk > drink

DrakeYang committed:
calculateMoney함수 이동: main > Vendingmachine.
메인함수에 수정내용 적용

DrakeYang committed:
함수명 변경 checkInventoryCount > getInventoryDetail.
리턴값이 InventoryDetail 이고 재고정보의 개수체크는 내부에서 이루어짐

DrakeYang committed:
함수명 변경 getOrderCount > receiveOrderCount

DrakeYang committed:
checkEnoughDrinkCount 함수 삭제. 주문수량과 재고수량을 비교하는건 음료이동 함수에 있는 기능임

DrakeYang committed:
음료주문 결과 조건문 이동:main > OutputView

DrakeYang committed:
재고수량과 주문수량 비교기능을 재고정보 내부로 옮김.
수정된 기능 메인함수에 적용

DrakeYang committed:
함수명 변경 : checkEnoughDrinkCount > isEnoughDrinkCount

DrakeYang commented:
피드백 부탁드립니다!

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -9,22 +9,175 @@
+        else {
+            outputView.printMessage(message: OutputView.errorMessage.wrongMenu.rawValue)
+        }
+    }
```
여기 while 반복문말고 위에 있는 함수들은 main 이 아니라 별도 객체로 만들수 있을까요?
지금 단계에서 말고 다음 단계를 진행하면서 위에 있는 함수들도 리팩토링을 해보세요.

godrm commented:
다음 단계로 넘어가서 계속해봅시다.
